### PR TITLE
added loading rle to grid, tests for the methods, and glider gui

### PR DESCRIPTION
**Pull request by NMandel527**

### CODE REVIEW

File: src/main/java/mandel/rle/Main.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package mandel.rle;
+
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+public class Main {
+    public static void main(String[] args) {
+        try {
+            String filePath = "C:/Users/necha/Downloads/file.rle";
```
Comment (andrewoid) [inline src/main/java/mandel/rle/Main.java]: Never. Ever. Ever. Put a path into git that only works on your machine.

File: src/main/java/mandel/rle/Rle.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+               count = 0;
+            }
+        }
+    }
+
+    public int[][] getGrid() {
+        return grid;
+    }
+
+    public void nextGen()
```
Comment (andrewoid) [inline src/main/java/mandel/rle/Rle.java]: You already wrote this code once. Why did you copy it?

File: src/main/java/mandel/rle/RleComponent.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
+package mandel.rle;
+
+import javax.swing.*;
+import java.awt.*;
+
+public class RleComponent extends JComponent {
```
Comment (andrewoid) [inline src/main/java/mandel/rle/RleComponent.java]: You already have a Component that displays a Grid, why create another one?

File: src/main/java/mandel/rle/RleFrame.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package mandel.rle;
+
+import javax.swing.*;
+
+public class RleFrame extends JFrame {
```
Comment (andrewoid) [inline src/main/java/mandel/rle/RleFrame.java]: Why did you create a new Frame?

File: src/test/java/mandel/rle/RleTest.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        assertEquals(1, grid[2][2]);
+
+        assertEquals(0, grid[0][0]);
+        assertEquals(0, grid[0][2]);
+        assertEquals(0, grid[1][1]);
+        assertEquals(0, grid[3][3]);
+    }
+
+    @Test
+    public void nextGen() {
```
Comment (andrewoid) [inline src/test/java/mandel/rle/RleTest.java]: You have a test for this already...

File: README.md [hunk 6]
```diff
@@ -23,4 +23,20 @@ and has a GUI that allows the user to play the Game of Life.
 
 [Conway's Game of Life Wikipedia Page](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life)
 
-[Play Conway's Game of Life](https://playgameoflife.com/)
\ No newline at end of file
+[Play Conway's Game of Life](https://playgameoflife.com/)
+
+
+## <br>RLE
```
Comment (andrewoid) [inline README.md]: Seems strange to put a br tag here.

File: src/main/java/mandel/rle/Rle.java (hunk 2 above)
Comment (NMandel527) [reply src/main/java/mandel/rle/Rle.java]: I didn't want to delete all the code that I had written from the last assignment, so I added the nextGen() to this class and that's why I also added another frame and component. Should I delete the code from last week and overwrite the component and the frame so it only shows the glider when I run it?

Comment (andrewoid) [reply src/main/java/mandel/rle/Rle.java]: You don't need to delete any code. 

You can add the code to your existing code. 
You don't need a new frame or component.

Comment (NMandel527) [reply src/main/java/mandel/rle/Rle.java]: Oh ok that makes sense - I get it now, thank you

File: src/main/java/mandel/gameoflife/GridComponent.java [hunk 7]
```diff
@@ -32,14 +32,24 @@ public void mouseClicked(MouseEvent e) {
     public void paintComponent(Graphics g) {
         super.paintComponent(g);
 
+        super.paintComponent(g);
```
Comment (andrewoid) [inline src/main/java/mandel/gameoflife/GridComponent.java]: Extra line

File: src/main/java/mandel/gameoflife/GridFrame.java [hunk 8]
```diff
@@ -4,21 +4,22 @@
...
+    public GridFrame(Grid grid) {
+        this.grid = grid;
         setSize(1000, 800);
         setTitle("Game of Life");
         setDefaultCloseOperation(EXIT_ON_CLOSE);
 
-        GridComponent gridComponent = new GridComponent(grid);
-        add(gridComponent, BorderLayout.CENTER);
+        GridComponent gliderComponent = new GridComponent(grid);
+        add(gliderComponent, BorderLayout.CENTER);
```
Comment (andrewoid) [inline src/main/java/mandel/gameoflife/GridFrame.java]: This name doesn't make sense

File: src/main/java/mandel/gameoflife/Main.java [hunk 9]
```diff
@@ -1,8 +1,34 @@
...
+        try {
+            String filePath = "https://conwaylife.com/patterns/glider.rle";
+
+            URL url = new URL(filePath);
+            BufferedReader reader = new BufferedReader(new InputStreamReader(url.openStream()));
+            StringBuilder content = new StringBuilder();
+            String line;
+            while ((line = reader.readLine()) != null) {
+                content.append(line).append("\n");
+            }
```
Comment (andrewoid) [inline src/main/java/mandel/gameoflife/Main.java]: What does this do?

File: src/test/java/mandel/gameoflife/GridTest.java [hunk 10]
```diff
@@ -6,6 +6,8 @@
 
 class GridTest
 {
+    Grid rle = new Grid(10, 10);
```
Comment (andrewoid) [inline src/test/java/mandel/gameoflife/GridTest.java]: Why is this here?

File: src/test/java/mandel/gameoflife/GridTest.java [hunk 11]
```diff
@@ -23,4 +25,48 @@ public void nextGen()
...
+        assertEquals(1, grid[2][2]);
+
+        assertEquals(0, grid[0][0]);
+        assertEquals(0, grid[0][2]);
+        assertEquals(0, grid[1][1]);
+        assertEquals(0, grid[3][3]);
+    }
+
+    @Test
+    public void decodeRle() {
```
Comment (andrewoid) [inline src/test/java/mandel/gameoflife/GridTest.java]: Technically this is getting tested twice since it is tested by the previous test

File: src/main/java/mandel/gameoflife/Grid.java [hunk 12]
```diff
@@ -11,6 +11,56 @@ public Grid(int width, int height) {
...
+                if (count == 0) {
+                    count = 1;
+                }
+                switch (c) {
+                    case 'b':
+                        col += count;
+                        break;
+                    case 'o':
+                        for (int j = 0; j < count; j++) {
+                            grid[row][col] = 1;
```
Comment (andrewoid) [inline src/main/java/mandel/gameoflife/Grid.java]: What if this goes out of bounds?

File: src/main/java/mandel/gameoflife/GridComponent.java [hunk 13]
```diff
@@ -6,12 +6,13 @@
 import java.awt.event.MouseEvent;
 
 public class GridComponent extends JComponent {
-
     private final Grid grid;
     private final int cellSize = 20;
 
     public GridComponent(Grid grid) {
         this.grid = grid;
+        grid.setInitial();
```
Comment (andrewoid) [inline src/main/java/mandel/gameoflife/GridComponent.java:14]: So this method only needs to be called once?

File: src/main/java/mandel/gameoflife/Main.java (hunk 9 above)
Comment (NMandel527) [reply src/main/java/mandel/gameoflife/Main.java]: It continuously reads a line from the input file (the rle link) as long as the line is not null. Then it appends the line to a StringBuilder. At the end of the loop, the StringBuilder should contain all the lines from the file. I just moved this into the loadFromRle method because it seemed more efficient and I adjusted the method and tests accordingly.

Comment (andrewoid) [reply src/main/java/mandel/gameoflife/Main.java]: Why don't you just read that whole file all at once?

Comment (NMandel527) [reply src/main/java/mandel/gameoflife/Main.java]: How can I read the whole file at once while still going through it line by line to get the information for the grid? Or is there a better way to get the information from the file without having to go through it line by line?

Comment (andrewoid) [reply src/main/java/mandel/gameoflife/Main.java]: Funny. The documentation on this isn't great. Don't worry about it. Let's discuss it in class

Comment (NMandel527) [reply src/main/java/mandel/gameoflife/Main.java]: Ok thank you so much. Is there anything else I should change/fix in my code?

Comment (andrewoid) [reply src/main/java/mandel/gameoflife/Main.java]: In left a number of other comments

Comment (NMandel527) [reply src/main/java/mandel/gameoflife/Main.java]: I made some changes and pushed the new code. It looks like it went through on my end, but should I push it again?

**Review by andrewoid — APPROVED** [review]
